modules/types/include/hephaestus/types/bounds.h

```diff
@@ -0,0 +1,100 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <concepts>
+#include <cstdint>
+#include <iostream>
+#include <random>
+
+#include <fmt/ostream.h>
+
+#include "hephaestus/random/random_object_creator.h"
+#include "hephaestus/utils/concepts.h"
+#include "hephaestus/utils/exception.h"
+
+namespace heph::types {
+
+enum class BoundsType : uint8_t {
+  INCLUSIVE,   // []
+  LEFT_OPEN,   // (]
+  RIGHT_OPEN,  // [)
+  OPEN         // ()
+};
+
+/// @brief A class representing range bounds.
+template <NumericType T>
+struct Bounds {
+  [[nodiscard]] inline constexpr auto operator==(const Bounds&) const -> bool = default;
+
+  [[nodiscard]] static auto random(std::mt19937_64& mt) -> Bounds;
+
+  [[nodiscard]] inline constexpr auto isWithinBounds(T value) const -> bool;
+
+  T lower_bound{};
+  T upper_bound{};
+  BoundsType bounds_type{ BoundsType::INCLUSIVE };
+};
+
+template <NumericType T>
+auto operator<<(std::ostream& os, const Bounds<T>& bounds) -> std::ostream&;
```

Comment:
rather provide a `format_as` instead of this?

Reply:
We keep ostream operator and fmt specialization, as fmt alone would not allow to read google tests output.

Reply:
fine with me, formatter is just much harder to read as `format_as` while achieving the same.
For google tests, you can do 
EXPECT_EQ(asdf) << fmt::format(..);

but yes thats no so nice. Alternatively you could define `operator<<` just using `fmt::format` in there.